dialogue-core/src/main/java/com/palantir/dialogue/core/CautiousIncreaseAggressiveDecreaseConcurrencyLimiter.java

```diff
@@ -69,17 +69,28 @@
      * ignore/dropped/success depending on the success or failure state of the response.
      * */
     Optional<Permit> acquire() {
-        int currentInFlight = getInflight();
-
-        // We don't want to hand out a permit if there are 4 inflight and a limit of 4.1, as this will immediately send
-        // our inflight number to 5, which is clearly above the limit.  Instead, we wait until there is capacity for
-        // one whole request before handing out a permit. In the worst-case scenario with zero inflight and a limit of
-        // 1, we'll still hand out a permit
-        if (currentInFlight <= getLimit() - 1) {
-            int inFlightSnapshot = inFlight.incrementAndGet();
-            return Optional.of(new Permit(inFlightSnapshot));
+
+        // Capture the limit field reference once to avoid work in a tight loop. The JIT cannot
+        // reliably optimize out references to final fields due to the potential for reflective
+        // modification.
+        AtomicInteger localInFlight = inFlight;
+        AtomicDouble localLimit = limit;
+        while (true) {
+
+            // We don't want to hand out a permit if there are 4 inflight and a limit of 4.1, as this will immediately
+            // send our inflight number to 5, which is clearly above the limit.
+            // Instead, we wait until there is capacity for one whole request before handing out a permit.
+            // In the worst-case scenario with zero inflight and a limit of 1, we'll still hand out a permit.
+            int currentInFlight = localInFlight.get();
+            if (currentInFlight >= localLimit.get()) {
```

Comment:
I think we should snapshot the limit value (including the cast to an integer) before we enter this loop. Either way there's an allowable data-race because the limit may change between the time it's read, and when we verify currentInFight, but that's not problematic.

Reply:
This comparison isn't the inverse of the original `currentInFlight <= getLimit() - 1` because values are converted to `double` rather than `integer`.

Reply:
I guess it's a slightly larger window for inconsistency this way, but yeah why not.